esx_service/vmdk_ops.py

```diff
@@ -777,31 +776,90 @@ def disk_attach(vmdk_path, vm):
         except vim.fault.VimFault as ex:
             msg=("Failed to add PVSCSI Controller: %s", ex.msg)
             return err(msg)
+            
         # Find the controller just added
         devices = vm.config.hardware.device
         pvsci = [d for d in devices
                  if type(d) == vim.ParaVirtualSCSIController and
                  d.key == controller_key]
-        pci_slot_number = get_controller_pci_slot(vm, pvsci,
+        pci_slot_number = get_controller_pci_slot(vm, pvsci[0],
                                                   offset_from_bus_number)
 
     # Find a slot on the controller, issue attach task and wait for completion
     if not disk_slot:
-        taken = set([dev.unitNumber
+        idx = 0
+           
+        while ((disk_slot is None) and (idx < len(pvsci))):
```

Comment:
This code needs a little refactoring so that logic isn't duplicated.

1. If there are pvscsi controllers already in the VM then start this loop and figure if we can find a free disk slot across all of the pvscsi controllers.

2. If either no pvscsi controllers or no free disk slots (len(pvscsi) >= 0) then

3. Add a pvscsi controller and use a disk slot in the newly added pvscsi controller.

That way there is no need to add a pvscsi controller above (if len(pvscsi) == 0) and then below again have logic to add one more pvscsi controller if we are out of disk slots.